DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs

```diff
@@ -0,0 +1,207 @@
+        {
+            new CourseStatisticsWithAdminFieldResponseCounts
+            {
+                ApplicationName = "Course",
+                CustomisationName = "Customisation",
+                Active = true,
+                CourseTopic = "Topic 1",
+                CategoryName = "Category 1",
+                HideInLearnerPortal = true,
+                DelegateCount = 1,
+                CompletedCount = 1,
+            },
+        };
+
+        private readonly List<Category> categories = new List<Category>
+        {
+            new Category { CategoryName = "Category 1" },
+            new Category { CategoryName = "Category 2" },
+        };
+
+        private readonly List<Topic> topics = new List<Topic>
+        {
+            new Topic { CourseTopic = "Topic 1" },
+            new Topic { CourseTopic = "Topic 2" },
+        };
+
+        private DelegateCoursesController controller = null!;
+        private DelegateCoursesController controllerWithCookies = null!;
+        private ICourseCategoriesDataService courseCategoryDataService = null!;
+        private ICourseService courseService = null!;
+        private ICourseTopicsDataService courseTopicsDataService = null!;
+        private HttpRequest httpRequest = null!;
+        private HttpResponse httpResponse = null!;
+
+        [SetUp]
+        public void Setup()
+        {
+            courseCategoryDataService = A.Fake<ICourseCategoriesDataService>();
+            courseTopicsDataService = A.Fake<ICourseTopicsDataService>();
+            courseService = A.Fake<ICourseService>();
+
+            A.CallTo(() => courseService.GetCentreSpecificCourseStatisticsWithAdminFieldResponseCounts(A<int>._, A<int>._)).Returns(courses);
+            A.CallTo(() => courseCategoryDataService.GetCategoriesForCentreAndCentrallyManagedCourses(A<int>._))
+                .Returns(categories);
+            A.CallTo(() => courseTopicsDataService.GetCourseTopicsAvailableAtCentre(A<int>._)).Returns(topics);
+            A.CallTo(
+                () => courseService.GetApplicationOptionsAlphabeticalListForCentre(A<int>._, A<int?>._, A<int?>._)
+            ).Returns(applicationOptions);
+
+            httpRequest = A.Fake<HttpRequest>();
+            httpResponse = A.Fake<HttpResponse>();
+
+            controller = new DelegateCoursesController(
+                courseService,
+                courseCategoryDataService,
+                courseTopicsDataService
+            )
+            .WithDefaultContext()
+            .WithMockUser(true, 101)
+            .WithMockTempData();
+
+            const string cookieName = "DelegateCoursesFilter";
+            const string cookieValue = "Status|Active|false";
+
+            controllerWithCookies = new DelegateCoursesController(
+                courseService,
+                courseCategoryDataService,
+                courseTopicsDataService
+            )
+            .WithMockHttpContext(httpRequest, cookieName, cookieValue, httpResponse)
```

Comment:
This may not work depending on how the FakeItEasy implements mocks, but can we pull the cookie-setting part of this method out into a separate extension method so we can set cookies in test code rather than setup code?
It's always a bit harder to read a test when some of the important values are set up in a different bit of code. Particularly the first test is a bit confusing.

Reply:
I think I initially set this up ages ago, and remember it being a bit of a pain. I get the feeling we do need to set it up this way, but can't remember exactly why. If we can get away with it, then extension methods might be nice. Otherwise, we can move the creation of the `controllerWithCookies` into the tests that need it. Though if we do end up needing to create it in multiple locations, then creating it here is probably for the best.